### ISSUE TITLE
Extend no-else-return to include process.exit

### ISSUE THREAD

**Issue by Nokel81** [eslint/eslint#10029]
Currently `no-else-return` does not look for `process.exit()` calls. However, these have a similar effect to that of a `return` when within a function (from the point of view of control flow. So it would make sense to also look for these

**What rule do you want to change?**
`no-else-return`

**Does this change cause the rule to produce more or fewer warnings?**
`more`

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option, `includeProcessExit` which will extend the lint to watch for calls to that function

**Please provide some example code that this change will affect:**

```js
// incorrect use with `includeProcessExit: true`
function (arg1) {
    if (arg1 == true) {
        process.exit(1);
    } else {
        console.log("All good");
    }
}

// correct use with `includeProcessExit: true`
function (arg1) {
    if (arg1 == true) {
        process.exit(1);
    }

    console.log("All good");
}
```

**What does the rule currently do for this code?**
Nothing

**What will the rule do after it's changed?**
Match with `process.exit()` calls as well as `return` statements

**Comment by Nokel81** [comment]
@platinumazure Is that better?

**Comment by platinumazure** [comment]
@Nokel81 Much better, thank you! And thank you for the issue.

I don't know if we'll consider making a core change in ESLint since we try not to put in runtime-specific features into core rules. (Yes, I'm aware we have a bunch of rules that are Node-focused now, but we wouldn't let those in if someone proposed those as core rules now.)

I would be very curious to know if we could augment our [code path analysis](https://eslint.org/docs/developer-guide/code-path-analysis) to allow developers (maybe parser developers?) to define what special function calls might represent a function termination, and thus allow *all* rules which use code path analysis to work consistently. But, I'm not sure the implementation and maintenance effort would be worth the benefits here.

I think your best option, at this point, is to create a custom rule. You could use [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer) to pull in the current `no-else-return` rule and then add other lint messages for cases where a `process.exit()` is inside an `else` clause.

**Comment by Nokel81** [comment]
Thank you for the response, I will see what I can do with the rule-composer. However, the item isn't about

```js
if (...) {
   ...
} else {
   return;
}
```

That would be another rule that would be nice to have but I think that should be its own issue

**Comment by not-an-aardvark** [comment]
I think `eslint-plugin-node` already has a similar rule [here](https://github.com/mysticatea/eslint-plugin-node/blob/5260039c77fd7e6368c1a5808f90462e5beec6f3/lib/rules/process-exit-as-throw.js), although it relies on modifying some of ESLint's internal state.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
